pkg/resources/rbac.authorization.k8s.io/v1/role/validator.go

```diff
@@ -0,0 +1,117 @@
+	objectsv1 "github.com/rancher/webhook/pkg/generated/objects/rbac.authorization.k8s.io/v1"
+	admissionv1 "k8s.io/api/admission/v1"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	v1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/utils/trace"
+)
+
+const (
+	grOwnerLabel = "authz.management.cattle.io/gr-owner"
+)
+
+// Validator implements admission.ValidatingAdmissionHandler.
+type Validator struct {
+	admitter admitter
+}
+
+// NewValidator returns a new validator for roles.
+func NewValidator() *Validator {
+	return &Validator{
+		admitter: admitter{},
+	}
+}
+
+// GVR returns the GroupVersionKind for this CRD.
+func (v *Validator) GVR() schema.GroupVersionResource {
+	return schema.GroupVersionResource{
+		Group:    "rbac.authorization.k8s.io",
+		Version:  "v1",
+		Resource: "roles",
+	}
+}
+
+// Operations returns list of operations handled by this validator.
+func (v *Validator) Operations() []admissionregistrationv1.OperationType {
+	return []admissionregistrationv1.OperationType{
+		admissionregistrationv1.Update,
+	}
+}
+
+// ValidatingWebhook returns the ValidatingWebhook used for this CRD.
+func (v *Validator) ValidatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) []admissionregistrationv1.ValidatingWebhook {
+	webhook := admission.NewDefaultValidatingWebhook(v, clientConfig, admissionregistrationv1.NamespacedScope, v.Operations())
+	webhook.ObjectSelector = &metav1.LabelSelector{
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			{
+				Key:      grOwnerLabel,
+				Operator: metav1.LabelSelectorOpExists,
+			},
+		},
+	}
+	return []admissionregistrationv1.ValidatingWebhook{*webhook}
+}
+
+// Admitters returns the admitter objects used to validate roles.
+func (v *Validator) Admitters() []admission.Admitter {
+	return []admission.Admitter{&v.admitter}
+}
+
+type admitter struct{}
+
+// Admit is the entrypoint for the validator. Admit will return an error if it's unable to process the request.
+// If this function is called without NewValidator(..) calls will panic.
+func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResponse, error) {
+	listTrace := trace.New("roleValidator Admit", trace.Field{Key: "user", Value: request.UserInfo.Username})
+	defer listTrace.LogIfLong(admission.SlowTraceDuration)
+
+	oldRole, newRole, err := objectsv1.RoleOldAndNewFromRequest(&request.AdmissionRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	fldPath := field.NewPath("role")
+	fieldErr := validateGROwnerLabel(oldRole, newRole, fldPath)
+	if fieldErr != nil {
+		return admission.ResponseBadRequest(fieldErr.Error()), nil
+	}
+
+	return admission.ResponseAllowed(), nil
+}
+
+// validateGROwnerLabel checks that a user isn't adding, removing or modifying the grOwnerLabel.
+// Users can add the label if it didn't already exist.
```

Comment:
I'm confused by the following:

> validateGROwnerLabel checks that a user isn't adding [...] the grOwnerLabel

and 

> Users can add the label if it didn't already exist.

Can user add the label or not?

Reply:
Good catch. I originally was going to prevent adding the label but opted to allow it because my GR changes in rancher/rancher will update the Role to add the label.

See here: https://github.com/rancher/rancher/blob/3284ace2d964abbf79a370a64eaabf8ecccda664/pkg/controllers/management/auth/globalroles/globalrole_handler.go#L300C1-L300C30